File: 1. Creational Design Patterns/Abstract Factory/C#/AbstractFactoryExample.cs
// Abstract Factory Pattern Example in C#
// Scenario: Cross-platform UI library that can create Windows and Mac style Buttons & Checkboxes.
// No external libraries are required. Compile with: csc AbstractFactoryExample.cs && AbstractFactoryExample.exe

using System;

namespace AbstractFactoryDemo
{
    // Product A
    interface IButton
    {
        void Paint();
    }

    // Product B
    interface ICheckbox
    {
        void Paint();
    }

    // Concrete Product A1
    class WindowsButton : IButton
    {
        public void Paint() => Console.WriteLine("Render a button in Windows style.");
    }

    // Concrete Product B1
    class WindowsCheckbox : ICheckbox
    {
        public void Paint() => Console.WriteLine("Render a checkbox in Windows style.");
    }

    // Concrete Product A2
    class MacButton : IButton
    {
        public void Paint() => Console.WriteLine("Render a button in macOS style.");
    }

    // Concrete Product B2
    class MacCheckbox : ICheckbox
    {
        public void Paint() => Console.WriteLine("Render a checkbox in macOS style.");
    }

    // Abstract Factory
    interface IGUIFactory
    {
        IButton CreateButton();
        ICheckbox CreateCheckbox();
    }

    // Concrete Factory 1
    class WindowsFactory : IGUIFactory
    {
        public IButton CreateButton() => new WindowsButton();
        public ICheckbox CreateCheckbox() => new WindowsCheckbox();
    }

    // Concrete Factory 2
    class MacFactory : IGUIFactory
    {
        public IButton CreateButton() => new MacButton();
        public ICheckbox CreateCheckbox() => new MacCheckbox();
    }

    // Client
    class Application
    {
        private readonly IButton _button;
        private readonly ICheckbox _checkbox;

        public Application(IGUIFactory factory)
        {
            _button = factory.CreateButton();
            _checkbox = factory.CreateCheckbox();
        }

        public void Paint()
        {
            _button.Paint();
            _checkbox.Paint();
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            IGUIFactory factory;
            if (args.Length > 0 && args[0].Equals("mac", StringComparison.OrdinalIgnoreCase))
                factory = new MacFactory();
            else
                factory = new WindowsFactory();

            var app = new Application(factory);
            app.Paint();
        }
    }
}

File: 1. Creational Design Patterns/Abstract Factory/Javascript/AbstractFactoryExample.js
// Abstract Factory Pattern – JavaScript Example
// Usage: node AbstractFactoryExample.js [mac]

// Product A
class Button {
  paint() {
    throw new Error('paint() must be implemented');
  }
}

// Product B
class Checkbox {
  paint() {
    throw new Error('paint() must be implemented');
  }
}

// Concrete Product A1
class WindowsButton extends Button {
  paint() {
    console.log('Render a button in Windows style.');
  }
}

// Concrete Product B1
class WindowsCheckbox extends Checkbox {
  paint() {
    console.log('Render a checkbox in Windows style.');
  }
}

// Concrete Product A2
class MacButton extends Button {
  paint() {
    console.log('Render a button in macOS style.');
  }
}

// Concrete Product B2
class MacCheckbox extends Checkbox {
  paint() {
    console.log('Render a checkbox in macOS style.');
  }
}

// Abstract Factory
class GUIFactory {
  createButton() {}
  createCheckbox() {}
}

// Concrete Factory 1
class WindowsFactory extends GUIFactory {
  createButton() {
    return new WindowsButton();
  }
  createCheckbox() {
    return new WindowsCheckbox();
  }
}

// Concrete Factory 2
class MacFactory extends GUIFactory {
  createButton() {
    return new MacButton();
  }
  createCheckbox() {
    return new MacCheckbox();
  }
}

// Client
class Application {
  constructor(factory) {
    this.button = factory.createButton();
    this.checkbox = factory.createCheckbox();
  }
  paint() {
    this.button.paint();
    this.checkbox.paint();
  }
}

const factory = process.argv[2] && process.argv[2].toLowerCase() === 'mac'
  ? new MacFactory()
  : new WindowsFactory();

const app = new Application(factory);
app.paint();
